modules/programs/firefox.nix

```diff
@@ -233,37 +257,78 @@ in {
             };
 
             bookmarks = mkOption {
-              type = types.attrsOf (types.submodule ({ config, name, ... }: {
-                options = {
-                  name = mkOption {
-                    type = types.str;
-                    default = name;
-                    description = "Bookmark name.";
-                  };
-
-                  keyword = mkOption {
-                    type = types.nullOr types.str;
-                    default = null;
-                    description = "Bookmark search keyword.";
+              type = let
+                bookmarkSubmodule = types.submodule ({ config, name, ... }: {
+                  options = {
+                    name = mkOption {
+                      type = types.str;
+                      default = name;
+                      description = "Bookmark name.";
+                    };
+
+                    keyword = mkOption {
+                      type = types.nullOr types.str;
+                      default = null;
+                      description = "Bookmark search keyword.";
+                    };
+
+                    url = mkOption {
+                      type = types.str;
+                      description = "Bookmark url, use %s for search terms.";
+                    };
                   };
+                }) // {
+                  description = "bookmark submodule";
+                };
 
-                  url = mkOption {
-                    type = types.str;
-                    description = "Bookmark url, use %s for search terms.";
+                bookmarkType = types.addCheck bookmarkSubmodule (x: x ? "url");
+
+                directoryType = types.submodule ({ config, name, ... }: {
+                  options = {
+                    name = mkOption {
+                      type = types.str;
+                      default = name;
+                      description = "Directory name.";
+                    };
+
+                    bookmarks = mkOption {
+                      type = types.listOf bookmarkType;
+                      default = [ ];
+                      description = "Bookmarks within directory.";
+                    };
                   };
+                }) // {
+                  description = "directory submodule";
                 };
-              }));
-              default = { };
+              in with types;
+              either (attrsOf bookmarkType)
```

Comment:
Would it be feasible to continue using an attribute set and distinguish between bookmarks and folders by inspecting their attributes (instead of using a list) when the user wishes to use both? That seems to involve less migration, but I might be missing something.

Reply:
I tried doing that in the start, but attribute sets does not keep any order. We could still model it as an attribute set and let users use the home-manager DAG functionality, but it seems painful if they want to keep a custom order for all of the bookmarks.